File: src/jaslib_src/jaslib/language_services/janome_ex/word_extraction/text_location.py
from __future__ import annotations

from typing import TYPE_CHECKING, final, override

from autoslot import Slots
from jaspythonutils.sysutils.weak_ref import WeakRef, WeakRefable

from jaslib import app

if TYPE_CHECKING:
    from jaslib.language_services.janome_ex.tokenizing.analysis_token import IAnalysisToken
    from jaslib.language_services.janome_ex.word_extraction.candidate_word_variant import CandidateWordVariant
    from jaslib.language_services.janome_ex.word_extraction.text_analysis import TextAnalysis

from jaspythonutils.sysutils.ex_str import newline

from jaslib.language_services.janome_ex.word_extraction.candidate_word import CandidateWord

_max_lookahead = 12  # In my collection the longest so far is 9, so 12 seems a pretty good choice.

@final
class TextAnalysisLocation(WeakRefable, Slots):
    def __init__(self, analysis: WeakRef[TextAnalysis], token: IAnalysisToken, character_start_index: int, token_index: int) -> None:
        self.weakref = WeakRef(self)
        self.next: WeakRef[TextAnalysisLocation] | None = None
        self.previous: WeakRef[TextAnalysisLocation] | None = None
        self.token: IAnalysisToken = token
        self.is_shadowed_by: list[WeakRef[TextAnalysisLocation]] = []
        self.shadows: list[WeakRef[TextAnalysisLocation]] = []
        self.analysis: WeakRef[TextAnalysis] = analysis
        self.token_index: int = token_index
        self.character_start_index: int = character_start_index
        self.character_end_index: int = character_start_index + len(self.token.surface) - 1

        self.display_variants: list[CandidateWordVariant] = []
        self.indexing_variants: list[CandidateWordVariant] = []
        self.candidate_words: list[CandidateWord] = []
        self.display_words: list[CandidateWord] = []

    @override
    def __repr__(self) -> str:
        return f"""
TextLocation('{self.character_start_index}-{self.character_end_index}, {self.token.surface} | {self.token.base_form})
{newline.join([cand.__repr__() for cand in self.indexing_variants])}
"""

    def forward_list(self, length: int = 99999) -> list[TextAnalysisLocation]:
        return self.analysis().locations[self.token_index: self.token_index + length + 1]

    @property
    def non_compound_candidate(self) -> CandidateWord: return self.candidate_words[-1]

    def analysis_step_1_analyze_non_compound_validity(self) -> None:
        lookahead_max = min(_max_lookahead, len(self.forward_list(_max_lookahead)))
        self.candidate_words = [CandidateWord([location.weakref for location in self.forward_list(index)]) for index in range(lookahead_max - 1, -1, -1)]
        self.candidate_words[-1].run_validity_analysis()  # the non-compound part needs to be completed first

    def analysis_step_2_analyze_compound_validity(self) -> None:
        for range_ in self.candidate_words[:-1]:  # we already have the last one completed
            range_.run_validity_analysis()

        self.indexing_variants = [variant for cand in self.candidate_words for variant in cand.indexing_variants]

    def run_display_analysis_and_update_display_words_pass_true_if_there_were_changes(self) -> bool:
        changes_made = False
        for range_ in self.candidate_words:
            if range_.run_display_analysis_pass_true_if_there_were_changes():
                changes_made = True

        if changes_made:
            self.display_words = [it for it in self.candidate_words if it.display_variants]
        return changes_made

    def analysis_step_3_run_display_analysis_without_shadowing_information_so_that_all_valid_matches_are_displayed_and_can_be_accounted_for_in_yielding_to_upcoming_compounds(self) -> None:
        self.run_display_analysis_and_update_display_words_pass_true_if_there_were_changes()

    def analysis_step_4_set_initial_shadowing_and_recalculate_display_words_return_true_on_changes(self) -> bool:
        display_words_updated = self.run_display_analysis_and_update_display_words_pass_true_if_there_were_changes()
        self.update_shadowing()
        return display_words_updated

    def analysis_step_5_update_shadowing_and_recalculate_display_words_return_true_on_changes(self) -> bool:
        display_words_updated = self.run_display_analysis_and_update_display_words_pass_true_if_there_were_changes()
        if display_words_updated:
            self.update_shadowing()
        return display_words_updated

    def update_shadowing(self) -> None:
        if self.display_words and not self.display_words[0].is_shadowed:
            self.display_variants = self.display_words[0].display_variants
            covering_forward_count = self.display_words[0].location_count - 1
            for shadowed in self.forward_list(covering_forward_count)[1:]:
                shadowed.is_shadowed_by.append(self.weakref)
                self.shadows.append(shadowed.weakref)
                shadowed._clear_shadowed()
        else:
            self._clear_shadowed()

    def _clear_shadowed(self) -> None:
        for shadowed_shadowed in self.shadows:
            shadowed_shadowed().is_shadowed_by.remove(self.weakref)
        self.shadows.clear()

    def is_next_location_inflecting_word(self) -> bool:
        return self.next is not None and self.next().is_inflecting_word()

    # todo: having this check here only means that marking a compound as an inflecting word has no effect, and figuring out why things are not working can be quite a pain
    def is_inflecting_word(self) -> bool:
        vocab = app.col().vocab.with_any_form_in([self.token.base_form, self.token.surface])
        return any(voc for voc in vocab if voc.matching_configuration.bool_flags.is_inflecting_word.is_active)

File: src/src_dotnet/JAStudio.Core/LanguageServices/JanomeEx/WordExtraction/TextLocation.cs
using System.Collections.Generic;
using System.Linq;
using JAStudio.Core.LanguageServices.JanomeEx.Tokenizing;

namespace JAStudio.Core.LanguageServices.JanomeEx.WordExtraction;

public sealed class TextAnalysisLocation
{
   const int MaxLookahead = 12; // In my collection the longest so far is 9, so 12 seems a pretty good choice.

    public TextAnalysisLocation? Next { get; set; }
    public TextAnalysisLocation? Previous { get; set; }
    public IAnalysisToken Token { get; }
    public List<TextAnalysisLocation> IsShadowedBy { get; }
    public List<TextAnalysisLocation> Shadows { get; }
    public TextAnalysis Analysis { get; }
    public int TokenIndex { get; }
    public int CharacterStartIndex { get; }
    public int CharacterEndIndex { get; }

    public List<CandidateWordVariant> DisplayVariants { get; set; }
    public List<CandidateWordVariant> IndexingVariants { get; set; }
    public List<CandidateWord> CandidateWords { get; set; }
    public List<CandidateWord> DisplayWords { get; set; }

    public TextAnalysisLocation(TextAnalysis analysis, IAnalysisToken token, int characterStartIndex, int tokenIndex)
    {
        Token = token;
        IsShadowedBy = [];
        Shadows = [];
        Analysis = analysis;
        TokenIndex = tokenIndex;
        CharacterStartIndex = characterStartIndex;
        CharacterEndIndex = characterStartIndex + token.Surface.Length - 1;

        DisplayVariants = [];
        IndexingVariants = [];
        CandidateWords = [];
        DisplayWords = [];
    }

    public override string ToString()
    {
        var variants = string.Join("\n", IndexingVariants.Select(cand => cand.ToString()));
        return $@"
TextLocation('{CharacterStartIndex}-{CharacterEndIndex}, {Token.Surface} | {Token.BaseForm})
{variants}
";
    }

    public List<TextAnalysisLocation> ForwardList(int length = 99999)
    {
        var actualLength = System.Math.Min(length + 1, Analysis.Locations.Count - TokenIndex);
        return Analysis.Locations.GetRange(TokenIndex, actualLength);
    }

    public CandidateWord NonCompoundCandidate => CandidateWords[^1];

    public void AnalysisStep1AnalyzeNonCompoundValidity()
    {
        var lookaheadMax = System.Math.Min(MaxLookahead, ForwardList(MaxLookahead).Count);
        CandidateWords = Enumerable.Range(0, lookaheadMax)
            .Reverse()
            .Select(index => new CandidateWord(ForwardList(index)))
            .ToList();
        CandidateWords[^1].RunValidityAnalysis(); // the non-compound part needs to be completed first
    }

    public void AnalysisStep2AnalyzeCompoundValidity()
    {
        foreach (var range in CandidateWords.Take(CandidateWords.Count - 1)) // we already have the last one completed
        {
            range.RunValidityAnalysis();
        }

        IndexingVariants = CandidateWords
            .SelectMany(cand => cand.IndexingVariants)
            .ToList();
    }

    public bool RunDisplayAnalysisAndUpdateDisplayWordsPassTrueIfThereWereChanges()
    {
        var changesMade = false;
        foreach (var range in CandidateWords)
        {
            if (range.RunDisplayAnalysisPassTrueIfThereWereChanges())
            {
                changesMade = true;
            }
        }

        if (changesMade)
        {
            DisplayWords = CandidateWords.Where(it => it.DisplayVariants.Any()).ToList();
        }
        return changesMade;
    }

    public void AnalysisStep3RunDisplayAnalysisWithoutShadowingInformationSoThatAllValidMatchesAreDisplayedAndCanBeAccountedForInYieldingToUpcomingCompounds()
    {
        RunDisplayAnalysisAndUpdateDisplayWordsPassTrueIfThereWereChanges();
    }

    public bool AnalysisStep4SetInitialShadowingAndRecalculateDisplayWordsReturnTrueOnChanges()
    {
        var displayWordsUpdated = RunDisplayAnalysisAndUpdateDisplayWordsPassTrueIfThereWereChanges();
        UpdateShadowing();
        return displayWordsUpdated;
    }

    public bool AnalysisStep5UpdateShadowingAndRecalculateDisplayWordsReturnTrueOnChanges()
    {
        var displayWordsUpdated = RunDisplayAnalysisAndUpdateDisplayWordsPassTrueIfThereWereChanges();
        if (displayWordsUpdated)
        {
            UpdateShadowing();
        }
        return displayWordsUpdated;
    }

    public void UpdateShadowing()
    {
        if (DisplayWords.Any() && !DisplayWords[0].IsShadowed)
        {
            DisplayVariants = DisplayWords[0].DisplayVariants;
            var coveringForwardCount = DisplayWords[0].LocationCount - 1;
            foreach (var shadowed in ForwardList(coveringForwardCount).Skip(1))
            {
                shadowed.IsShadowedBy.Add(this);
                Shadows.Add(shadowed);
                shadowed.ClearShadowed();
            }
        }
        else
        {
            ClearShadowed();
        }
    }

    void ClearShadowed()
    {
        foreach (var shadowedShadowed in Shadows)
        {
            shadowedShadowed.IsShadowedBy.Remove(this);
        }
        Shadows.Clear();
    }

    public bool IsNextLocationInflectingWord() => Next != null && Next.IsInflectingWord();

    // todo: having this check here only means that marking a compound as an inflecting word has no effect, and figuring out why things are not working can be quite a pain
    public bool IsInflectingWord()
    {
        var vocab = TemporaryServiceCollection.Instance.App.Col().Vocab.WithAnyFormIn([Token.BaseForm, Token.Surface]);
        return vocab.Any(voc => voc.MatchingConfiguration.BoolFlags.IsInflectingWord.IsActive);
    }
}
